### ISSUE TITLE
Canonical way to solve spurious protection rchk error with destructors?

### ISSUE THREAD

**Issue by vspinu** [kalibera/cran-checks#5]
I think I have a classical issue of a spurious warning as mentioned in the official [docs](https://github.com/kalibera/cran-checks/blob/master/rchk/PROTECT.md#hints-for-interpreting-the-reports) . How would I go about silencing the rchk CRAN message?

My rchk warning looks like this:

```
Function Unnester::process(SEXPREC*, SEXPREC*)
  [UP] unprotected variable <unnamed var:   %4 = alloca %struct.SEXPREC*, align 8> while calling allocating function NodeAccumulator::~NodeAccumulator() unnest/src/./unnest.h:284
```

The error comes from this [block](https://github.com/vspinu/unnest/blob/9d42d65b690238b9799987dd66483f764d6c0934/src/unnest.h#L281-L283), and while I understand the reasoning I don't quite see how to fix it elegantly. 

What I came up with is this workaround:

```
      SEXP out;
      {
        NodeAccumulator acc;
        add_node(*this, acc, vacc, NilSpec, spec, 0, x);
        out = PROTECT(build_df(acc, vacc));
      }
      UNPROTECT(1);
      return out;
```
Is this a recommended way to silent the warning? If so, it might be useful to add it to the official doc page. For guys like me the solution is not immediately apparent. 

Thanks!

**Comment by vspinu** [comment]
Actually just assigning the output of a function locally removes the warning:

```c++
      SEXP out = build_df(acc, vacc);
      return out;
```
This doesn't quite fit into my understanding. `out` is still unprotected, and if the destructor allocates it can remove `out` from the heap. So why is there no warning in this case?

**Comment by kalibera** [comment]
I think that in principle we should be modifying our code to make the tool happy only when it also makes the code cleaner or easier to read (or at least not harder). Otherwise I think it is better to ignore the warning, when we know for sure it is a false alarm, adding a note to the code, why. Of course, it is a matter of opinion to say whether this change explicitly invoking the NodeAccumulator via {} is making the code cleaner/easier to read or not - maybe it is, see below, but maybe it could be encapsulated somewhere, and my preference would be adding a comment there that this is to protect "out" against the destructor of "NodeAccumulator".

In this case, when I am looking only at this fragment, but not the rest of the code (which actually may be a good benchmark of what is easy to read). Then I am asking myself why do I have to create the NodeAccumulator locally, when I am only passing it once to add_node, but don't need it anymore. I start speculating whether it defines some memory for add_node to work in, and if it is the case, whether it is actually safe to use the result of build_df after the NodeAccumulator is destroyed. I assume it is ok, otherwise the code would have been broken before, but then - then perhaps this should be encapsulated into a variant of build_df, so that the intricacies of calling the destructor explicitly is at least hidden.

I probably don't know a better trick to protect a value from a destructor.

**Comment by kalibera** [comment]
> Actually just assigning the output of a function locally removes the warning:
> 
> ```c++
>       SEXP out = build_df(acc, vacc);
>       return out;
> ```
> 
> This doesn't quite fit into my understanding. `out` is still unprotected, and if the destructor allocates it can remove `out` from the heap. So why is there no warning in this case?

Yes, this makes no difference for correctness, but for some reason the tool is confused by this.

**Comment by vspinu** [comment]
> I am asking myself why do I have to create the NodeAccumulator locally, when I am only passing it once to add_node, but don't need it anymore

NodeAccumulator indeed can be constructed inside build_df, but the second argument VarAccumulator cannot be, so for some symetry sake both are passed as arguments. But, now that you pointed it out, I think it makes sense to hide it in. 

> Yes, this makes no difference for correctness, but for some reason the tool is confused by this.

Got it, I think I started seeing the big picture :) 

> Otherwise I think it is better to ignore the warning, when we know for sure it is a false alarm, adding a note to the code, why.

Sure, but CRAN does not ignore it, and each time I would need to manually confirm that the warning is ok. Or is there an magic comment that would inhibit the warning?

**Comment by kalibera** [comment]
> Sure, but CRAN does not ignore it, and each time I would need to manually confirm that the warning is ok. Or is there an magic comment that would inhibit the warning?

Yes, that's right, the decision when it still makes sense to update the code in a CRAN package to make sure rchk is happy is more difficult than when developing just code for yourself. Unfortunately most of the cases that get to me when CRAN package authors claim that certain warnings are false alarms turn out to be true errors, so supporting pragmas would not work. There is a filter I apply on the reports, you can see the difference between "results" and "results-full", the latter is the vanilla output from `rchk`. So I can be persuaded in exceptional cases to filter out e.g. a complete package from the checks, particularly for packages using C++, and I filter out some cases when the tool knows it is not so sure about its warnings. Even this sometimes will hide true errors, but when the false alarms rate is too high, that's necessary.

In the past I've implemented scripts to look at diffs - to find out new reports, assuming the reports found earlier were false alarms. I used it only for base R at that time. It could probably be resurrected in principle, but would require some work, especially gluing it to somehow differential checks.
